### PR TITLE
First working prototype - DHCPv4

### PR DESCRIPTION
**Pull request by yoelcaspersen**
This version works with kernel 5.15 and LLVM 13.

Improvements:

- Interface name is added to Option 82 Circuit ID
- On server replies, Option 82 is wiped completely
- Char array parameters replaced with structs

Functions that previously used char arrays as function parameters now
use structs instead - otherwise the verifier cannot perform proper
boundary checks.

To-do:
- IPv6 support
- Multiple interface support

### CODE REVIEW

File: dhcp-relay/Makefile [hunk 1]
```diff
@@ -4,6 +4,7 @@ USER_TARGETS := dhcp_user_xdp
 BPF_TARGETS :=dhcp_kern_xdp 
 EXTRA_DEPS := dhcp-relay.h
 #EXTRA_CFLAGS := $(if $(IPV6),-DIPV6)
+EXTRA_CFLAGS := -fno-builtin
```
Comment (tohojo) [inline dhcp-relay/Makefile:7]: It would be better if we could build without this...

File: dhcp-relay/dhcp-relay.h [hunk 2]
```diff
@@ -10,19 +10,21 @@
 #define DHO_DHCP_AGENT_OPTIONS 82
 #define RAI_CIRCUIT_ID 1
 #define RAI_REMOTE_ID 2
-#define RAI_OPTION_LEN 40
+#define RAI_OPTION_LEN IF_NAMESIZE
```
Comment (tohojo) [inline dhcp-relay/dhcp-relay.h:13]: Is this really long enough? Theoretically the interface itself can be IF_NAMESIZE long?

File: dhcp-relay/dhcp_kern_xdp.c [hunk 3]
```diff
@@ -49,76 +56,104 @@ struct {
 	__uint(max_entries, 16384);
 } client_vlans SEC(".maps");
 
+void memcpy_var(void *to, void *from, __u64 len) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: With MAX_LOOPS this can fail, so it should probably no longer be void, and instead return an error code if it does...

File: dhcp-relay/dhcp_kern_xdp.c [hunk 4]
```diff
@@ -49,76 +56,104 @@ struct {
...
+	__u8 *t8 = to, *f8 = from;
+	int i;
+
+	for (i = 0; i < len && i < MAX_LOOPS; i++) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: I think it should be enough to just add a check for len > MAX_LOOPS (and return error if it is)?

File: dhcp-relay/dhcp_kern_xdp.c [hunk 5]
```diff
@@ -49,76 +56,104 @@ struct {
...
+		*t8++ = *f8++;		
+	}
+		
+}
+
+void memset_var(void *d, __u8 c, __u64 len) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: Same as above

File: dhcp-relay/dhcp_kern_xdp.c [hunk 6]
```diff
@@ -49,76 +56,104 @@ struct {
...
+		
+}
+
 /* Inserts DHCP option 82 into the received DHCP packet
  * at the specified offset.
  */
 static __always_inline int write_dhcp_option_82(void *ctx, int offset,
 	struct collect_vlans *vlans, char *dev) {
+	
 	struct dhcp_option_82 option;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: Why not just zero out the whole struct with an initialiser (`= {}` here) instead of the not-quite-_var memsets below?

Comment (tohojo) [reply dhcp-relay/dhcp_kern_xdp.c]: Or even initialise the options directly using {} syntax...

File: dhcp-relay/dhcp_kern_xdp.c [hunk 7]
```diff
@@ -49,76 +56,104 @@ struct {
...
-
+	
+	static __u8 buf[RAI_OPTION_LEN];
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: This can also be initialised to 0 instead of the memset below

File: dhcp-relay/dhcp_kern_xdp.c [hunk 8]
```diff
@@ -49,76 +56,104 @@ struct {
...
+	 * contains null bytes.
 	 */
-	 
-	char buf[IF_NAMESIZE];
-	memset(buf, '-', sizeof (buf));
+
+	memset(buf, 0, sizeof (buf));
+	
+	int c = VLAN_ASCII_MAX; /* We will need 4 bytes at most */
+	int i = RAI_OPTION_LEN - 1;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: sizeof(buf)?

File: dhcp-relay/dhcp_kern_xdp.c [hunk 9]
```diff
@@ -49,76 +56,104 @@ struct {
...
 	
-	int c = VLAN_ASCII_MAX;		/* We will need 4 bytes at most */
-	int i = IF_NAMESIZE - 1;
 	__u16 inner_vlan = vlans->id[1];
 	__u16 outer_vlan = vlans->id[0];
-	
+
+	/* Convert inner VLAN to ASCII */
+#pragma unroll VLAN_ASCII_MAX
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: Is this really needed?

File: dhcp-relay/dhcp_kern_xdp.c [hunk 10]
```diff
@@ -49,76 +56,104 @@ struct {
...
 	for (c = VLAN_ASCII_MAX; c > 0; c--) {
 		buf[i--] = (inner_vlan % 10) + '0';
 		inner_vlan /= 10;
 		if (inner_vlan == 0) {
 			break;
 		}
 	}
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: As I said elsewhere, this should be a helper function

File: dhcp-relay/dhcp_kern_xdp.c [hunk 11]
```diff
@@ -49,76 +56,104 @@ struct {
...
 
 	buf[i--] = '.';
-	
-	for (c = IF_NAMESIZE - 1; c >= 0; c--) {
 
-		if (dev[c] != 0) {
+	/* Append interface name */
+#pragma unroll RAI_OPTION_LEN
+	for (c = RAI_OPTION_LEN - 1; c >= 0; c--) {
+		if (dev[c] != 0)
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: `if (dev[c] == 0) break;` ?

File: dhcp-relay/dhcp_kern_xdp.c [hunk 12]
```diff
@@ -49,76 +56,104 @@ struct {
...
 	}
 	
-	/* Initialize remote ID */
-	memset(option.remote_id.val, 0, sizeof(option.remote_id.val));
-	option.remote_id.option_id = RAI_REMOTE_ID;
-	option.remote_id.len = sizeof(option.remote_id.val);
+	i++;
+	
+	/* Copy resulting interface name to circuit_id */
+	if (sizeof (option.circuit_id.val) == sizeof (buf)) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: Why this check?

File: dhcp-relay/dhcp_kern_xdp.c [hunk 13]
```diff
@@ -49,76 +56,104 @@ struct {
...
+		memcpy_var(option.circuit_id.val, buf + i, sizeof (buf) - i);
+		//memcpy_var(option.circuit_id.val, buf, sizeof (buf));
+	}
 	
 	return xdp_store_bytes(ctx, offset, &option, sizeof (option), 0);
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: The option should be resized to the right size based on the string length as well, no?

File: dhcp-relay/Makefile (hunk 1 above)
Comment (yoelcaspersen) [reply dhcp-relay/Makefile:7]: Sure, but then we'll need to find a way to copy an arbitrary number of bytes without having the compiler default to memcpy().

File: dhcp-relay/dhcp-relay.h (hunk 2 above)
Comment (yoelcaspersen) [reply dhcp-relay/dhcp-relay.h:13]: Right now I am just trying to make it work with an interface name that contains the VLAN tags as part of the name (and therefore is at most IF_NAMESIZE long). Once it works with 16 bytes, increasing to e.g. 32 or more should simply be a matter of increasing RAI_OPTION_LEN.

File: dhcp-relay/dhcp_kern_xdp.c (hunk 11 above)
Comment (yoelcaspersen) [reply dhcp-relay/dhcp_kern_xdp.c]: It is counting down - the first couple of bytes are more than likely to be null. It needs to skip the null bytes before copying the actual interface name.

File: dhcp-relay/dhcp_kern_xdp.c (hunk 12 above)
Comment (yoelcaspersen) [reply dhcp-relay/dhcp_kern_xdp.c]: The only way I could make it pass the verifier.

File: dhcp-relay/dhcp_kern_xdp.c (hunk 13 above)
Comment (yoelcaspersen) [reply dhcp-relay/dhcp_kern_xdp.c]: Yes, ideally it should, but as far as I can see, that would rule out the use of xdp_store_bytes(). Copying dynamic lengths of memory really is an uphill battle with XDP.

The current version actually works in the sense that it can send a packet to the DHCP server and receive a reply. But the verifier explodes if I disable the check that skips single tagged packets (like the reply that comes from the server):

```
	if (vlans.id[1] == 0) {
		bpf_printk("No inner VLAN tag set\n");
		goto out;
	}
```
The verifier complains that the BPF program is too large. I guess it cannot cope with the loops.

File: dhcp-relay/dhcp_kern_xdp.c (hunk 10 above)
Comment (yoelcaspersen) [reply dhcp-relay/dhcp_kern_xdp.c]: Will it decrease the resulting eBPF object size or make it easier to pass the verifier?

Bear in mind, convert an int between 0 and 4095 to ASCII is a very specific thing to do...

Comment (tohojo) [reply dhcp-relay/dhcp_kern_xdp.c]: ...but you're doing it twice? That alone merits a helper :)

As for decreasing complexity, yes if the functions are *not* inlined, the verifier should be able to do per-function verification, which can decrease overall complexity. Probably not worth it for this helper, but the full write_option82 helper could be a contender for this.

File: dhcp-relay/dhcp_kern_xdp.c (hunk 13 above)
Comment (tohojo) [reply dhcp-relay/dhcp_kern_xdp.c]: My guess would be that this check allows the compiler to make some assumptions that changes the code flow, and removing it makes things longer. Or something along those lines. The above-mentioned per-function verification might be a way around this...

File: dhcp-relay/dhcp_kern_xdp.c (hunk 10 above)
Comment (yoelcaspersen) [reply dhcp-relay/dhcp_kern_xdp.c]: Helper function u16_to_ascii(char *buf, __u8 offset, __u16 num) added now.

File: dhcp-relay/dhcp_kern_xdp.c (hunk 13 above)
Comment (yoelcaspersen) [reply dhcp-relay/dhcp_kern_xdp.c]: Apparently the primary issue was use of char arrays as functions parameters. After an upgrade to kernel 5.15 I replaced the char arrays with pointers to structs containing the necessary char arrays. 

That seems to work, as long as just the right amount of boundary checks are littered across the code.

File: dhcp-relay/dhcp-relay.h [hunk 14]
```diff
@@ -38,6 +43,11 @@ struct dhcp_option_255 {
 	__u8 t;
 };
 
+struct dev_name {
+	char name[IF_NAMESIZE];
+	__u8 len;
```
Comment (tohojo) [inline dhcp-relay/dhcp-relay.h]: len is unneeded, see below

File: dhcp-relay/dhcp_kern_xdp.c [hunk 15]
```diff
@@ -49,77 +56,162 @@ struct {
...
+//
+//	}
+//
+//
+//	if (i == RAI_OPTION_LEN)
+//		return -1;
+//
+//	return 0;
+//
+//}
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: Please remove commented-out code

File: dhcp-relay/dhcp_kern_xdp.c [hunk 16]
```diff
@@ -49,77 +56,162 @@ struct {
...
+
+int u16_to_ascii(struct sub_option *opt, __u8 offset, __u16 num) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: This is q fairly small helper function to make global; better keep this 'static' so it can be inlined, and make the whole `write_dhcp_option_82` function below global. Just having these helpers global are not enough to pass the verifier on my kernel either.

Also, since there are two different functions with an 'offset' parameter, but with different naming, rename the 'offset' parameter to 'end_offset'.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 17]
```diff
@@ -49,77 +56,162 @@ struct {
...
+
+		opt->val[i] = (__u8) (num % 10) + '0';
+		num /= 10;
+		if (num == 0)
+			break;
+
+	}
+
+	if (i > 0) {
+		opt->val[--i] = '.';
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: The function is called `u16_to_ascii`, so what is this doing here? Better leave this to the caller.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 18]
```diff
@@ -49,77 +56,162 @@ struct {
...
+	}
+
+	if (i > RAI_OPTION_LEN)
+		return -1;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: This check is not needed...

File: dhcp-relay/dhcp_kern_xdp.c [hunk 19]
```diff
@@ -49,77 +56,162 @@ struct {
...
+	__u8 i;
+	//#pragma unroll U16_ASCII_LEN
+	for (i = offset - 1; i > 0; i--) {
+
+		if (i == 0)
+			break;
+
+		opt->val[i] = (__u8) (num % 10) + '0';
+		num /= 10;
+		if (num == 0)
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: Right now the value of `i` is different depending on which condition exited the loop: if this one matches, `i` will be the last character written, but if the loop condition itself matches, `i` will be right before the last character written (i.e., -1). The code below assumes this is always the condition that matches (which it will be, in practice), but that is not clear from reading the code; so better to move this check into the loop condition as well, so `i` always ends up right before the last character written. This also requires setting `i = offset` above instead of `i = offset - 1`, as then things fit when passing `i` back in for multiple calls.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 20]
```diff
@@ -49,77 +56,162 @@ struct {
...
+	}
+
+	if (i > RAI_OPTION_LEN)
+		return -1;
+
+	return i;
+
+}
+
+int str_len(struct dev_name *dev) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: This is not needed either, see below

File: dhcp-relay/dhcp_kern_xdp.c [hunk 21]
```diff
@@ -49,77 +56,162 @@ struct {
...
+	for (i = 0; i < RAI_OPTION_LEN; i++) {
+
+		if (i < dev->len) {
+			/* Copy device name */
+			dest->val[i] = dev->name[i];
+		} else if (offset < RAI_OPTION_LEN) {
+			/* Move VLAN part (all bytes from offset and up) */
+			dest->val[i] = dest->val[offset];
+			dest->val[offset++] = 0;
+		}
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: This weird double-construction supposes a particular construction of the buffer being written to, which is hard to understand and makes the code less readable (and less reusable as well). Instead, how about turning this function into one that actually does what the name applies, like:

```c
static int copy_dev_name(struct sub_option *dest, __u8 wr_offset,
			 struct dev_name *dev, __u8 rd_offset) {
	int i;

	if (dest == NULL || dev == NULL || wr_offset > sizeof(dest->val) -1)
		return -1;

	for (i = 0; i < sizeof(dest->val) - wr_offset; i++) {
		if (i + rd_offset > sizeof(dev->name) ||
		    !dev->name[i + rd_offset])
			break;

		dest->val[i + wr_offset] = dev->name[i + rd_offset];
	}

 	return i;
}
```

and calling it like:

```c
	len = copy_dev_name(&option.circuit_id, 0, dev, 0);
	if (len < 0)
		return len;

	len = copy_dev_name(&option.circuit_id, len, &buf, vlan_offset);
	if (len < 0)
		return len;
```

where `buf` is a separate `struct dev_name` allocated on the stack, into which the VLAN tags are stringified.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 22]
```diff
@@ -49,77 +56,162 @@ struct {
...
+
+	}
+
+	return offset;
+}
+
 /* Inserts DHCP option 82 into the received DHCP packet
  * at the specified offset.
  */
 static __always_inline int write_dhcp_option_82(void *ctx, int offset,
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: As mentioned above, better to make this one the global function; that requires turning 'dev' into a pointer argument (pass-by-value is not supported). I'd also suggest renaming `offset` to `pkt_offset`, as we already have way too many `offset`s in this code...

File: dhcp-relay/dhcp_kern_xdp.c [hunk 23]
```diff
@@ -49,77 +56,162 @@ struct {
...
+	 * contains null bytes.
 	 */
-	 
-	char buf[IF_NAMESIZE];
-	memset(buf, '-', sizeof (buf));
-	
-	int c = VLAN_ASCII_MAX;		/* We will need 4 bytes at most */
-	int i = IF_NAMESIZE - 1;
+
+	int i = RAI_OPTION_LEN;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: The kernel style guidelines forbid mixing variable declarations and code, so please move the declarations to the top of the function. Also, `i` is a terrible variable name (only use that for idiomatic loops), maybe `vlan_offset`? Finally, you can use `sizeof(buf.name)` (and changing the `u16_to_ascii` calls to this `struct dev_name buf` as mentioned above.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 24]
```diff
@@ -49,77 +56,162 @@ struct {
...
-	
-	buf[i--] = '.';
-	
-	for (c = VLAN_ASCII_MAX; c > 0; c--) {
-		buf[i--] = (outer_vlan % 10) + '0';
-		outer_vlan /= 10;
-		if (outer_vlan == 0) {
-			break;
+
+	if (inner_vlan != 0) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: Everything seems to assume there are actually two VLAN tags present, so probably better to error out if either of the tags are unset, instead of this conditional (same below, of course).

File: dhcp-relay/dhcp_kern_xdp.c [hunk 25]
```diff
@@ -49,77 +56,162 @@ struct {
...
-		memcpy(option.circuit_id.val, buf, sizeof(buf));
-	}
-	
-	/* Initialize remote ID */
-	memset(option.remote_id.val, 0, sizeof(option.remote_id.val));
-	option.remote_id.option_id = RAI_REMOTE_ID;
-	option.remote_id.len = sizeof(option.remote_id.val);
-	
+	/* Insert device name and left-align circuit ID */
+	i = copy_dev_name(&option.circuit_id, i, &dev);
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: See above for this

File: dhcp-relay/dhcp_kern_xdp.c [hunk 26]
```diff
@@ -136,10 +228,10 @@ static __always_inline int write_dhcp_option_255(void *ctx, int offset) {
 
 /* Calculates the IP checksum */
 static __always_inline int calc_ip_csum(struct iphdr *oldip, struct iphdr *ip,
-	__u32 oldcsum) {
+		__u32 oldcsum) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:161]: Some odd alignment stuff going on here; generally, for multi-line function definitions, align the continuation of the parameters with the opening parenthesis on subsequent lines.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 27]
```diff
@@ -166,27 +258,18 @@ static __always_inline int calc_ip_csum(struct iphdr *oldip, struct iphdr *ip,
...
 int xdp_dhcp_relay(struct xdp_md *ctx) {
 
-	bpf_printk("\n");
-
 	/* Tail extend packet */
 	int res = bpf_xdp_adjust_tail(ctx, delta);
 	if (res != 0) {
 		bpf_printk("Cannot tail extend packet, delta %i - error code %i", delta, res);
-		return XDP_ABORTED;
+		return XDP_PASS;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:199]: This is a packet we figured we should handle, but that failed; so what is XDP_PASS supposed to do with it? Better to drop...

File: dhcp-relay/dhcp_kern_xdp.c [hunk 28]
```diff
@@ -217,34 +302,38 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 	int key = 0;
 	int len = 0;
 
-	if (data + 1 > data_end)
-		return XDP_ABORTED;
+	if (data + 1 > data_end) {
+		goto out;
+	}
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:237]: Unneded braces

File: dhcp-relay/dhcp_kern_xdp.c [hunk 29]
```diff
@@ -217,34 +302,38 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
 	if (ether_type < 0) {
-		bpf_printk("Cannot determine ethertype");
-		rc = XDP_ABORTED;
+		bpf_printk("Cannot determine EtherType");
 		goto out;
 	}
+
 	if (ether_type != bpf_htons(ETH_P_IP)) {
-		bpf_printk("Ethertype %#x is not ETH_P_IP", bpf_ntohs(ether_type));
+		//bpf_printk("Ethertype %x is not ETH_P_IP", bpf_ntohs(ether_type));
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:248]: Remove commented-out code...

File: dhcp-relay/dhcp_kern_xdp.c [hunk 30]
```diff
@@ -256,15 +345,19 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
 		goto out;
+	}
 
 	/* Increase IP length header */
 	ip->tot_len += bpf_htons(delta);
 
 	/* Increase UDP length header */
 	udp->len += bpf_htons(delta);
 
+	udp->check = 0;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:289]: Here you're modifying the packet, but any `goto out` statements below will end up returning `XDP_PASS`, passing up a potentially partially-modified packet. `rc` should be changed to `XDP_DROP` at the point of the first packet modification.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 31]
```diff
@@ -285,23 +378,20 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
-	if (dev == NULL)
+	dev_config = bpf_map_lookup_elem(&device_name, &key);
+	if (dev_config == NULL)
 		goto out;
 
-	/* Copy headers of packet to buf */
-	//if (xdp_load_bytes(ctx, 0, buf, static_offset))
-	//	goto out;
+	memcpy(dev.name, dev_config, RAI_OPTION_LEN);
+	dev.len = str_len(&dev);
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c]: This is just looping through the string to find the length of it, only to use that length as the end condition of a loop later. This is totally superfluous, just check for the null terminator directly in the loop that uses length and get rid of this call (and lose the `len` member of struct dev_name entirely).

File: dhcp-relay/dhcp_kern_xdp.c [hunk 32]
```diff
@@ -285,23 +378,20 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
 
 	/* Increment offset by 4 bytes for each VLAN (to accomodate VLAN headers */
+#pragma unroll VLAN_MAX_DEPTH
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:318]: Not needed

File: dhcp-relay/dhcp_kern_xdp.c [hunk 33]
```diff
@@ -317,6 +407,16 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 	}
 	dhcp = data + vlan_length + dhcp_offset;
 
+	/* Check hops */
+	if (dhcp->hops > 16) {
+		bpf_printk("Max hops exceeded, discarding packet");
+		rc = XDP_ABORTED;
+		goto out;
+	}
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:344]: Now we're suddenly doing parsing and sanity checks of the packet in the middle of modifying it. Better to parse and check everything first, *then* start modifying.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 34]
```diff
@@ -326,11 +426,11 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
-		&& eth->h_dest[1] == 0xff
-		&& eth->h_dest[2] == 0xff
-		&& eth->h_dest[3] == 0xff
-		&& eth->h_dest[4] == 0xff
-		&& eth->h_dest[5] == 0xff)) {
+			&& eth->h_dest[1] == 0xff
+			&& eth->h_dest[2] == 0xff
+			&& eth->h_dest[3] == 0xff
+			&& eth->h_dest[4] == 0xff
+			&& eth->h_dest[5] == 0xff)) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:362]: This calls for an `is_ether_broadcast()` helper (which can then be inverted below)

File: dhcp-relay/dhcp_kern_xdp.c [hunk 35]
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
+		/* Set source MAC */
+		memcpy(eth->h_source, relay_hwaddr, ETH_ALEN);
 
-	/* Increment hops */
-	dhcp->hops++;
+		/* Set destination IP */
+		ip->daddr = IP_ADDR_BCAST;
+
+		/* Set source IP */
+		ip->saddr = *relay_agent_ip;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:418]: Jumping back and forth between modifying L2 and L3 headers makes the code hard to follow. Also this function is way too long, it should be split: let's turn it into separate `handle_request()` and `handle_reply()` functions, and also split the VLAN tag insertion/removal into a separate helper.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 36]
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
+
+	}
 
 	/* Check if we exceed boundaries to make verifier happy */
 	if (data + offset > data_end)
 		goto out;
 
 	option_offset = offset;
 
+	__u8 n = 0;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:496]: See above cf mixing declarations and code. Also, like `i`, `n` is a terrible variable name outside very narrowly-scoped counter contexts.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 37]
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
+
 	__u8 *pos = (__u8 *) (data + option_offset);
 
 	/* Loop through all DHCP options */
-#pragma unroll DHCP_MAX_OPTIONS
+	//#pragma unroll DHCP_MAX_OPTIONS
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:501]: Remove

File: dhcp-relay/dhcp_kern_xdp.c [hunk 38]
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
 	for (i = 0; i < DHCP_MAX_OPTIONS; i++) {
 
 		/* Verifier check */
 		if (pos + 1 > data_end)
 			break;
 
+		/* Read option code */
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:508]: Redundant comment; it just says what the code is doing...

File: dhcp-relay/dhcp_kern_xdp.c [hunk 39]
```diff
@@ -357,74 +457,167 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
 		option_code = *pos;
 
 		bpf_printk("Got option code %i at offset %i, hex %x", option_code, option_offset, option_offset);
 
+		if (option_code == 82 && dhcp->op == DHCP_REPLY) {
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:513]: This code is wrong for several reasons:

- It doesn't handle option 82 for REQUESTS (it should drop the packet)
- It just blindly turns option 82 into an option 255, regardless of where it is in the packet (what if it's not actually the last option)
- It assumes that the option 82 is always the same size as the one we put in
- It leaves junk in the packet

The right thing to do is to shrink the packet; and either move everything up, including any non-255 options after this, or at least checking to see if there is anything other than an end option in the rest of the packet (and dropping it as invalid if there is).

File: dhcp-relay/dhcp_kern_xdp.c [hunk 40]
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
 
+		/* Verifier check */
+		if (pos + 1 > data_end) {
+			break;
+		}
+
 		option_length = *pos;
 		option_offset += option_length + 2;
 
+		/* Verifier check */
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:582]: These repeated 'verifier check' comments are just noise. There are also way too many checks one after another at the end here, some of them could definitely be collapsed. This may be easier to figure out after some of the other refactoring mentioned above.

File: dhcp-relay/dhcp_kern_xdp.c [hunk 41]
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
 		if (pos + 1 > data_end) {
 			break;
 		}
 		pos++;
 
+		/* Verifier check */
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:588]: For instance, the check below could be moved up, and changed to `if (pos + option_length + 1 > data_end)`

File: dhcp-relay/dhcp_kern_xdp.c [hunk 42]
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
-			if (vlans.id[i]) {*/
-	/*  */
-	/*if (xdp_store_bytes(ctx, offset, buf + offset,
-		4, 0))
-		return XDP_ABORTED;
-	offset += 4;
-}
+	/* Adjust IP offset for VLAN header when VLAN header has been added */
+	if (head_adjusted) {
+		ip = data + ip_offset + sizeof (struct vlan_hdr);
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:600]: Why carry this `head_adjusted` all the way down here instead of just modifying `ip_offset` directly when doing the adjustment?

File: dhcp-relay/dhcp_kern_xdp.c [hunk 43]
```diff
@@ -438,57 +631,50 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
...
+	} else {
+		ip = data + ip_offset;
 	}
-	}*/
-
-	ip = data + ip_offset;
-	if (ip + 1 > data_end)
+	if (ip + 1 > data_end) {
 		return XDP_ABORTED;
+	}
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:606]: Unnecessary braces

File: dhcp-relay/dhcp_kern_xdp.c [hunk 44]
```diff
@@ -499,7 +685,6 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 	/* Re-calculate ip checksum */
 	__u32 sum = calc_ip_csum(&oldip, ip, oldip.check);
 	ip->check = ~sum;
-	rc = XDP_PASS;
 
 	goto out;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:618]: Redundant goto

File: dhcp-relay/dhcp_kern_xdp.c [hunk 45]
```diff
@@ -499,7 +685,6 @@ int xdp_dhcp_relay(struct xdp_md *ctx) {
 	/* Re-calculate ip checksum */
 	__u32 sum = calc_ip_csum(&oldip, ip, oldip.check);
 	ip->check = ~sum;
-	rc = XDP_PASS;
```
Comment (tohojo) [inline dhcp-relay/dhcp_kern_xdp.c:502]: See above; this *is* actually the right place to change the return code back into something that sends out the packet, as this is where the packet is once again valid...

File: dhcp-relay/test.sh [hunk 46]
```diff
@@ -0,0 +1,65 @@
+#!/bin/bash -x
+
+OUTER_VLAN=83
+INNER_VLAN=20
+UPLINK_VLAN=84
+IF="ens6f0"
```
Comment (tohojo) [inline dhcp-relay/test.sh:6]: It should be possible to test this using network namespaces; having a test that is runnable without needing special hardware makes it way easier to test. Such a test should really be added so it can be run as part of a CI job.

**Review by tohojo — CHANGES_REQUESTED** [review]
Nice job getting something to work! Unfortunately the code has turned a wee bit spaghetti in the process; I guess most of the comments below could be summed up as changes that fix that :)

File: dhcp-relay/dhcp_kern_xdp.c (hunk 19 above)
Comment (tohojo) [reply dhcp-relay/dhcp_kern_xdp.c]: Actually, this difference is fine, as writing to the first character of the string without having `num` be 0 afterwards *should* be an error (i.e., -1 return); and moving the '.' assignment out of the function removes the assumption that this was the check that exited the loop. However, there's an off-by-one error in the loop condition above, it should be `i >= 0` (as writing to the first byte of the buffer is fine).

**Comment by tohojo** [comment]
Took the liberty of fixing some of the issues I pointed out myself, and pushed them to your branch :)
